packages/voting/src/reply.rs

```diff
@@ -0,0 +1,46 @@
+/// Masks for reply id
+const FAILED_PROPOSAL_EXECUTION_MASK: u64 = 0b00;
+const FAILED_PROPOSAL_HOOK_MASK: u64 = 0b01;
+const FAILED_VOTE_HOOK_MASK: u64 = 0b11;
+
+const BITS_RESERVED_FOR_REPLY_TYPE: u8 = 2;
+const REPLY_TYPE_MASK: u64 = (1 << BITS_RESERVED_FOR_REPLY_TYPE) - 1;
+
+/// Since we can only pass `id`, and we need to perform different actions in reply,
+/// we decided to take few bits to identify "Reply Type".
+/// See <https://github.com/DA0-DA0/dao-contracts/pull/385#discussion_r916324843>
+pub enum TaggedReplyId {
+    FailedProposalExecution(u64),
+    FailedProposalHook(u64),
+    FailedVoteHook(u64),
+}
+
+impl TaggedReplyId {
+    /// Takes `Reply.id` and returns tagged version of it,
+    /// depending on a first few bits.
+    ///
+    /// We know it costs extra pattern match, but cleaner code in `reply` Methods
+    pub fn new(id: u64) -> Self {
+        let reply_type = id & REPLY_TYPE_MASK;
+        let id = id >> BITS_RESERVED_FOR_REPLY_TYPE;
+        match reply_type {
+            FAILED_PROPOSAL_EXECUTION_MASK => TaggedReplyId::FailedProposalExecution(id),
+            FAILED_PROPOSAL_HOOK_MASK => TaggedReplyId::FailedProposalHook(id),
+            FAILED_VOTE_HOOK_MASK => TaggedReplyId::FailedVoteHook(id),
+            _ => unreachable!(),
```

Comment:
How would you feel about returning an error here?

Could make an error struct:

```rust
pub enum TagError {
  #[error("Unknown reply id ({id}).")]
  UnknownReplyId { id: u64 }
}
```

Then, in the contract error for our proposal modules we can add a variant like:

```rust
    #[error("{0}")]
   Tag(#[from] TagError),
```

And things like `?` will work between the two.

This purely a UX thing. Outcome for caller is the same.

Reply:
Not sure what we want to return in error, we have three fine options, I guess: 

1. `Reply.id`(pre-tagged)
2. `reply_type`(tag)
3. Both `reply_type` and id after bit shift

Reply:
Great! Thanks!